thegraph/src/subscriptions/ticket.rs

```diff
@@ -0,0 +1,416 @@
+//!
+//! See [`TicketClaims::to_verification_message`] method for the implementation details.
+
+use std::io::{Cursor, Write as _};
+
+pub use alloy_chains::{Chain, NamedChain};
+use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine as _};
+use ethers::signers::Wallet;
+use ethers_core::k256::ecdsa::SigningKey;
+use ethers_core::types::Signature;
+
+use crate::types::Address;
+
+/// Claims that are encoded in a ticket.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct TicketClaims {
+    /// Chain ID (EIP-155) of the chain on which the subscriptions contract is deployed.
+    pub chain_id: Chain,
+
+    /// Address of the subscriptions contract.
+    pub contract: Address,
+
+    /// Signer address that is authorized to sign the ticket.
+    ///
+    /// This address should be the user's address or one of the user's authorized signers.
+    pub signer: Address,
+
+    /// User address associated with the subscription.
+    ///
+    /// Required when the authorized `signer` is not the `user` associated with a subscription. When
+    /// omitted, the `signer` is implied to be equal to the `user`.
+    pub user: Option<Address>,
+
+    /// Optional name of the subscription.
+    pub name: Option<String>,
+
+    /// Comma-separated list of subgraphs that can be queried with this ticket.
+    pub allowed_subgraphs: Option<String>,
+
+    /// Comma-separated list of subgraph deployments that can be queried with this ticket.
+    pub allowed_deployments: Option<String>,
+
+    /// Comma-separated list of origin domains that can send queries with this ticket.
+    pub allowed_domains: Option<String>,
+}
+
+impl TicketClaims {
+    /// Returns the user address that the ticket is for.
+    ///
+    /// If the `user` field is not set, the `signer` is address is returned.
+    pub fn user(&self) -> Address {
+        self.user.unwrap_or(self.signer)
+    }
+
+    /// Returns the verification message that must be signed by the signer address.
+    ///
+    /// The verification message is a string that contains all the claims serialized in a
+    /// human-readable format. This serialized message is ready to be hashed and signed.
+    fn to_verification_message(&self) -> String {
+        let mut cursor = Cursor::<Vec<u8>>::default();
+
+        if let Some(allowed_deployments) = self.allowed_deployments.as_deref() {
+            writeln!(&mut cursor, "allowed_deployments: {}", allowed_deployments).unwrap();
+        }
+        if let Some(allowed_domains) = self.allowed_domains.as_deref() {
+            writeln!(&mut cursor, "allowed_domains: {}", allowed_domains).unwrap();
+        }
+        if let Some(allowed_subgraphs) = self.allowed_subgraphs.as_deref() {
+            writeln!(&mut cursor, "allowed_subgraphs: {}", allowed_subgraphs).unwrap();
+        }
+
+        writeln!(&mut cursor, "chain_id: {}", self.chain_id).unwrap();
```

Comment:
This doesn't actually serialize the chain ID. `self.chain_id` should be renamed to `self.chain` or the type should be `u64` to make it clear.

Reply:
We should add a unit test to check the `to_verification_message` output, to avoid nasty serialization problems.

Reply:
Added in https://github.com/edgeandnode/toolshed/commit/0ebc62e34707724f80c354c7cd6363ebe524c02c ✔️

Reply:
> This doesn't actually serialize the chain ID. `self.chain_id` should be renamed to `self.chain` or the type should be `u64` to make it clear.

I added the `#[serde_as(as = "serde_with::FromInto<u64>")]` to ensure that it is always serialized as a `u64`. 

If you prefer to rename the field to `chain`, I can add a field alias so _serde_cbor_2_ serializes the field with the `chain_id` tag.